### ISSUE TITLE
Clarify std::mem-docs wrt host/target

### ISSUE THREAD

**Issue by lukaslueg** [rust-lang/rust#81206]
Functions like `std::mem::align_of` (and also parts of `std::alloc::Layout`) which are const can be used to initialize other constants and, soon, types like `[u8; { std::mem::align_of::<String>() }]`. As far as I can see the docs currently do not explicitly make it clear that the compiler *guarantees* (including future versions) that the return values of those functions are valid for the _target_-platform and not for the _host_-platform, even in a const-context.

This is probably everybody's expectation. If this behavior is guaranteed, the docs should mention that, though.

**Comment by nagisa** [comment]
`host` vs `target` vs `build` is a concern build system needs to deal when selecting toolchains, libraries and flags to use. Such a distinction doesn't exist within the toolchain itself, however… if you ignore accidental breakage once in a while due to failures to normalize for byte order when handling the libraries, that is.

With that in mind I feel like documenting this on potentially every const function is excessive. It might make sense to put a line that states something along these lines in the reference somewhere, however.

**Comment by lukaslueg** [comment]
@nagisa I agree. My concern came up specifically with `std::mem::align_of`, where one might wonder if `const foo: usize = std::mem::align_of::<String>()` is String's alignment on the target-platform or it's alignment on the host-platform, which is definitely different.
